src/module.c

```diff
@@ -60,6 +68,63 @@ int _RegisterDataTypes(RedisModuleCtx *ctx) {
 	return REDISMODULE_OK;
 }
 
+static void _PrepareModuleGlobals() {
+	assert(pthread_mutex_init(&_module_mutex, NULL) == 0);
+	graphs_in_keyspace = array_new(GraphContext *, 1);
+}
+
+static void RG_ForkPrepare() {
+	/* At this point, a fork call has been issued. (We assume that this is because BGSave was called.)
+	 * Acquire the writer mutex of each graph to ensure that no locks are held, or else
+	 * the child process will deadlock when attempting to acquire that lock.
+	 * 1. If a writer thread is active, we'll wait until the writer finishes and releases the lock.
+	 * 2. Otherwise, no write in progress. We'll acquire the lock and release it immediately after forking. */
+
+	assert(pthread_mutex_lock(&_module_mutex) == 0); // Acquire the module-scoped lock.
+
+	uint graph_count = array_len(graphs_in_keyspace);
+	for(uint i = 0; i < graph_count; i++) {
+		// Acquire each writer mutex to guarantee that no graph is being modified.
+		Graph_WriterEnter(graphs_in_keyspace[i]->g);
+	}
+
+	assert(pthread_mutex_unlock(&_module_mutex) == 0); // Release the module-scoped lock.
```

Comment:
Not sure if we can release the module level lock here, what if a new graph is created right after this function terminates but before afterforkparent is called, or is there a guarantee that this can't happen

Reply:
Good call; I'll move this to the AfterFork routines!